InnerEye/ML/lightning_models.py

```diff
@@ -0,0 +1,980 @@
+    :return: A modified batch of data, where all tensor now live on the given CUDA device.
+    """
+    if not isinstance(batch, dict):
+        raise ValueError(f"This function expects a dictionary input, but got: {type(batch)}")
+    # For sequence models, this method receives a dictionary with "item": List[List[ScalarItem]]
+    items = batch.get("items", None)
+    if items is not None and isinstance(items, List) and isinstance(items[0], List) and \
+            isinstance(items[0][0], ScalarItem):
+        batch["items"] = [[j.to_device(device) for j in i] for i in items]
+        return batch
+    else:
+        return move_data_to_device(batch, device)
+
+
+def create_lightning_model(config: ModelConfigBase, set_optimizer_and_scheduler: bool = True) -> InnerEyeLightning:
+    """
+    Creates a PyTorch Lightning model that matches the provided InnerEye model configuration object.
+    The `optimizer` and `l_rate_scheduler` object of the Lightning model will also be populated.
+    :param set_optimizer_and_scheduler: If True (default), initialize the optimizer and LR scheduler of the model.
+    If False, skip that step (this is only meant to be used for unit tests.)
+    :param config: An InnerEye model configuration object
+    :return: A PyTorch Lightning model object.
+    """
+    if config.is_segmentation_model:
+        model: InnerEyeLightning = SegmentationLightning(config)
+    elif config.is_scalar_model:
+        model = ScalarLightning(config)
+    else:
+        raise NotImplementedError(f"Don't know how to handle config of type {type(config)}")
+    if set_optimizer_and_scheduler:
+        model.set_optimizer_and_scheduler(config)
+    return model
+
+
+def load_from_lightning_checkpoint(config: ModelConfigBase, checkpoint_path: Path) -> InnerEyeLightning:
+    """
+    Reads a PyTorch model from a checkpoint. First, a PyTorch Lightning model is created matching the InnerEye
+    model configuration, its parameter tensors are then populated from the given checkpoint.
+    :param config: An InnerEye model configuration object
+    :param checkpoint_path: The location of the checkpoint file.
+    :return: A PyTorch Lightning model object.
+    """
+    # Create a Lighting model that matches the configuration, but keep only the type of it
+    lightning_model_type = type(create_lightning_model(config))
+    # For model debugging, allow loading a GPU trained model onto the CPU. This will clearly only work
+    # if the model is small.
+    map_location = None if config.use_gpu else 'cpu'
+    lightning_model = lightning_model_type.load_from_checkpoint(checkpoint_path=str(checkpoint_path),
+                                                                map_location=map_location,
+                                                                config=config)
+    return lightning_model
+
+
+def adjust_model_for_inference(config: ModelConfigBase, lightning_model: InnerEyeLightning) -> None:
```

Comment:
From this, it looks like data parallel is only an option during train time, and model parallel an option during inference time, should we document this?

Reply:
Yes, that's a good idea. Can you suggest a good place? Or maybe add that in your docu update PR?

Reply:
I'll close this one here.